src/uct/cuda/cuda_copy/cuda_copy_md.c

```
@@ -551,66 +539,51 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
- * @param [out] base_address_p Returned base address
- * @param [out] alloc_length_p Returned size of the memory allocation
- *
- * @return Error code as defined by @ref ucs_status_t.
+ * @param [in]  cuda_device Device for which primary context is requested.
+ * @param [out] cuda_ctx_p  Returned context handle of the retained context.
  */
 static ucs_status_t
-uct_cuda_copy_md_get_address_range(const void *address, size_t length,
-                                   CUcontext ctx, void **base_address_p,
-                                   size_t *alloc_length_p)
+uct_cuda_copy_md_primary_ctx_retain(CUdevice cuda_device, CUcontext *cuda_ctx_p)
 {
+    unsigned int flags;
+    int active;
     ucs_status_t status;
-    CUdeviceptr base;
-    size_t alloc_length;
-    ucs_status_t status_ctx_pop;
-    CUcontext popped_ctx;
+    CUcontext cuda_ctx;
 
-    if (ctx != NULL) {
-        /* GetAddressRange requires context to be set. On DGXA100 it takes
-         * 0.03us to push and pop the context associated with address. */
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxPushCurrent(ctx));
-        if (status != UCS_OK) {
-            return status;
-        }
+    status = UCT_CUDADRV_FUNC_LOG_ERR(
+            cuDevicePrimaryCtxGetState(cuda_device, &flags, &active));
+    if (status != UCS_OK) {
+        return status;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
-            cuMemGetAddressRange(&base, &alloc_length, (CUdeviceptr)address));
-    if (ctx != NULL) {
-        status_ctx_pop = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxPopCurrent(&popped_ctx));
-        if (status != UCS_OK) {
-            /* cuMemGetAddressRange failed after pushing non-NULL context */
-            return UCS_ERR_INVALID_ADDR;
-        }
-
-        if (status_ctx_pop != UCS_OK) {
-            /* Failed to set the context that was current before the other
-             * context was pushed. */
-            return status_ctx_pop;
-        }
+    if (!active) {
```

do we need to check the state? Maybe we can just always retain? According to the doc:

Retains the primary context on the device. Once the user successfully retains the primary context, the primary context will be active and available to the user until the user releases it with [cuDevicePrimaryCtxRelease()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PRIMARY__CTX.html#group__CUDA__PRIMARY__CTX_1gf2a8bc16f8df0c88031f6a1ba3d6e8ad) or resets it with [cuDevicePrimaryCtxReset()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PRIMARY__CTX.html#group__CUDA__PRIMARY__CTX_1g5d38802e8600340283958a117466ce12). Unlike [cuCtxCreate()](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g65dc0012348bc84810e2103a40d8e2cf) the newly retained context is not pushed onto the stack.

---

I want to report an error if we try to retain a primary context that has not been activated yet.

---

isn't it errnous to do? Assume we have a primary context, but al the threads are currently using some other context. Then this check will fail, which is not what we want